### PR TITLE
sys/ztimer: add periph_ptp backend

### PR DESCRIPTION
**Pull request by Jnae**
### Contribution description

This PR adds support for PTP-Clock in ztimer. 

### Testing procedure

I did some basic tests on `nucleo-f767zi` of clock and timer and compared the time values printed with the shell timestamps. It requires to use `ztimer_usec`.  

### Issues/PRs references
None

### CODE REVIEW

File: sys/ztimer/periph_ptp.c [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+/**
+ * @ingroup     sys_ztimer_periph_ptp
+ * @{
+ *
+ * @file
+ * @brief       ztimer periph/ptp implementation
+ *
+ * @author      Jana Eisoldt <jana.eisoldt@ovgu.de>
+ 1
+ 1 @}
```
Comment (maribu) [inline sys/ztimer/periph_ptp.c:19]:
```suggestion
 *
 * @}
```

File: sys/ztimer/periph_ptp.c [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+static ztimer_clock_t *clock_timer;
+
+void ptp_timer_cb(void)
+{
+    ztimer_handler(clock_timer);
+}
+
+static void _ztimer_periph_ptp_set(ztimer_clock_t *clock, uint32_t val)
+{
+    (void) clock;
```
Comment (maribu) [inline sys/ztimer/periph_ptp.c:39]:
```suggestion
    (void)clock;
```

File: sys/ztimer/periph_ptp.c [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+    unsigned state = irq_disable();
+
+    clock_timer = clock;
+    ptp_timer_set_u64((uint64_t) val);
```
Comment (maribu) [inline sys/ztimer/periph_ptp.c:44]:
```suggestion
    ptp_timer_set_u64(val);
```

Integer will be automatically promoted to higher types if the signature of the function require this. The cast is only needed when reducing the range of the integer type to silence warnings from the compiler.

File: sys/ztimer/periph_ptp.c [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+    irq_restore(state);
+}
+
+static uint32_t _ztimer_periph_ptp_now(ztimer_clock_t *clock)
+{
+    (void)clock;
+    return (uint32_t) ptp_clock_read_u64();
```
Comment (maribu) [inline sys/ztimer/periph_ptp.c:52]:
```suggestion
    return (uint32_t)ptp_clock_read_u64();
```

**Review by maribu — COMMENTED** [review]
Some minor style nits, otherwise this looks good to me.

**Comment by maribu** [comment]
Please squash / amend directly.

**Comment by Jnae** [comment]
> Some minor style nits, otherwise this looks good to me.

Thank you, I changed it accordingly.

File: sys/ztimer/periph_ptp.c [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+static void _ztimer_periph_ptp_set(ztimer_clock_t *clock, uint32_t val)
+{
+    (void)clock;
+
+    unsigned state = irq_disable();
+
+    clock_timer = clock;
+    ptp_timer_set_u64(val);
```
Comment (kaspar030) [inline sys/ztimer/periph_ptp.c:44]: this needs to be underflow safe, if the underlying function is not! e.g., if setting 0, this needs to set a minimal value that will still trigger in the future.

**Comment by kaspar030** [comment]
Nice first contribution btw! :+1:

File: sys/ztimer/periph_ptp.c (hunk 5 above)
Comment (maribu) [reply sys/ztimer/periph_ptp.c:44]: This shouldn't be an issue for PTP, but it did make me realize that this will not work this easy: The PTP API assumes that the 64 bit value will never overflow. Hence, if one sets a timeout in the past, it will result in an IRQ right away. (A 64 bit period is just too long to be sensible. And even the hardware that I looked at implemented it this way, that the IRQ is triggered when the current value is greater than or equal to the target value. In fact, it is difficult to implement it differently in hardware: The actual clock frequency will be less than 1 GHz on all MCUs for quite some time, so the nanosecond counter (which is required by PTP as resolution) will always step more than one tick per clock cycle. Hence, only triggering an IRQ once the value is identical to the target value would likely result in the counter stepping over the target value without perfectly hitting it.)

However, if this is tested with a synchronized PTP clock, this will not work as it is: Internally, the time will already greater than `UINT32_MAX`, so this will always trigger an IRQ right away. The best would actually be to allow for 64 bit hardware clocks in ztimer. A quick hack would be to set the time like this:

```C
    uint64_t now = ptp_clock_read_u64();
    uint64_t target = (now & 0xffffffff00000000) | val;
    if (target < now) {
        target += 1ULL << 32;
    }
    ptp_timer_set_u64(target);
```

Comment (kaspar030) [reply sys/ztimer/periph_ptp.c:44]: Hm, for ztimer, this behavior is perfect. ztimer only ever sets relative values. I guess that's what `ptp_timer_set_u64()` expects? Otherwise there needs to be a read()/add()/set().

Comment (kaspar030) [reply sys/ztimer/periph_ptp.c:44]: (with "trigger in the future" I meant "trigger right away or in the future")

Comment (maribu) [reply sys/ztimer/periph_ptp.c:44]: > I guess that's what `ptp_timer_set_u64()` expects?

No, that is absolute. Looking at the other APIs, it would have made damn sense to name that `set_absolute()`. But I have another API change in the pipe anyway, so I could just also add `set_absolute()` and make this relative.

It is good that nobody uses the API so far (hopefully!)

Comment (maribu) [reply sys/ztimer/periph_ptp.c:44]: @Jnae so the conclusion is: the code here is good as it is, but you need to wait for me fixing the API.

Comment (maribu) [reply sys/ztimer/periph_ptp.c:44]: Wait, [`ptp_timer_set_u64()`](https://api.riot-os.org/group__drivers__periph__ptp.html#ga03812aa5adece70e82398147fce1d517) does expect relative times and there even is [`ptp_timer_set_absolute()`](https://api.riot-os.org/group__drivers__periph__ptp.html#ga7aca6b37425dc052ccf12a4bdf6b054a).

Why I'm always the person with the least knowledge about the code I have written?

So, conclusion is: The code is fine and this should work.

Comment (kaspar030) [reply sys/ztimer/periph_ptp.c:44]: > Why I'm always the person with the least knowledge about the code I have written?

that's why we team up. I'm usually the person with the least knowledge about the code I have written.

Comment (maribu) [reply sys/ztimer/periph_ptp.c:44]: Also: The test in `tests/periph_ptp_timer` includes testing for correct behavior with small relative sets (including a relative set of zero).

Comment (Jnae) [reply sys/ztimer/periph_ptp.c:44]: > Wait, [`ptp_timer_set_u64()`](https://api.riot-os.org/group__drivers__periph__ptp.html#ga03812aa5adece70e82398147fce1d517) does expect relative times and there even is [`ptp_timer_set_absolute()`](https://api.riot-os.org/group__drivers__periph__ptp.html#ga7aca6b37425dc052ccf12a4bdf6b054a).

Great, I was already doubting my skills of testing. Sorry that I didn't read your message until now, otherwise I would have told you.

File: sys/ztimer/periph_ptp.c [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+static const ztimer_ops_t _ztimer_periph_ptp_ops = {
+    .set = _ztimer_periph_ptp_set,
+    .now = _ztimer_periph_ptp_now,
+    .cancel = _ztimer_periph_ptp_cancel,
+};
+
+void ztimer_periph_ptp_init(ztimer_periph_ptp_t *clock)
+{
+    clock->ops = &_ztimer_periph_ptp_ops;
+    ptp_init();
```
Comment (maribu) [inline sys/ztimer/periph_ptp.c:70]:
```suggestion
```

`ptp_init()` is called as part of `periph_init()`. No need to initialize it again here.

File: sys/ztimer/periph_ptp.c [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+static const ztimer_ops_t _ztimer_periph_ptp_ops = {
+    .set = _ztimer_periph_ptp_set,
+    .now = _ztimer_periph_ptp_now,
+    .cancel = _ztimer_periph_ptp_cancel,
+};
+
+void ztimer_periph_ptp_init(ztimer_periph_ptp_t *clock)
+{
+    clock->ops = &_ztimer_periph_ptp_ops;
```
Comment (maribu) [inline sys/ztimer/periph_ptp.c:69]:
```suggestion
    clock->ops = &_ztimer_periph_ptp_ops;
    clock->max_value = UINT32_MAX;
    clock_timer = clock;
```

File: sys/ztimer/periph_ptp.c [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    ztimer_handler(clock_timer);
+}
+
+static void _ztimer_periph_ptp_set(ztimer_clock_t *clock, uint32_t val)
+{
+    (void)clock;
+
+    unsigned state = irq_disable();
+
+    clock_timer = clock;
```
Comment (maribu) [inline sys/ztimer/periph_ptp.c:43]:
```suggestion
```

This is better done once during initialization, see below.

File: sys/ztimer/periph_ptp.c [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+void ptp_timer_cb(void)
+{
+    ztimer_handler(clock_timer);
+}
+
+static void _ztimer_periph_ptp_set(ztimer_clock_t *clock, uint32_t val)
+{
+    (void)clock;
+
+    unsigned state = irq_disable();
```
Comment (maribu) [inline sys/ztimer/periph_ptp.c:41]:
```suggestion
```

Not needed, the PTP clock implementations are expected to disable IRQs, if needed. (And the only implementation we have does disable IRQs.)

File: sys/ztimer/periph_ptp.c [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+    clock_timer = clock;
+    ptp_timer_set_u64(val);
+
+    irq_restore(state);
```
Comment (maribu) [inline sys/ztimer/periph_ptp.c:46]:
```suggestion
```

**Review by maribu — COMMENTED** [review]
Some comments inline. This also exposed a bug in my timer implementation, I will PR the fix swiftly.

Could you also add this?

```diff
diff --git a/sys/ztimer/Makefile.dep b/sys/ztimer/Makefile.dep
index c72b3c3828..b8dfdebd19 100644
--- a/sys/ztimer/Makefile.dep
+++ b/sys/ztimer/Makefile.dep
@@ -70,6 +70,10 @@ ifneq (,$(filter ztimer_periph_rtt,$(USEMODULE)))
   FEATURES_REQUIRED += periph_rtt
 endif
 
+ifneq (,$(filter ztimer_periph_ptp,$(USEMODULE)))
+  FEATURES_REQUIRED += periph_ptp_timer
+endif
+
 ifneq (,$(filter ztimer_convert_frac,$(USEMODULE)))
   USEMODULE += frac
 endif
```

**Comment by maribu** [comment]
I used this for testing:

```diff
diff --git a/tests/ztimer_periodic/Makefile b/tests/ztimer_periodic/Makefile
index fa67ced1d8..062848c3c5 100644
--- a/tests/ztimer_periodic/Makefile
+++ b/tests/ztimer_periodic/Makefile
@@ -8,6 +8,6 @@ include ../Makefile.tests_common
 TEST_ON_CI_BLACKLIST += native
 
 USEMODULE += fmt
-USEMODULE += ztimer_usec ztimer_msec
+USEMODULE += ztimer_periph_ptp
 
 include $(RIOTBASE)/Makefile.include
diff --git a/tests/ztimer_periodic/main.c b/tests/ztimer_periodic/main.c
index 71673a52d7..34b9fd30ec 100644
--- a/tests/ztimer_periodic/main.c
+++ b/tests/ztimer_periodic/main.c
@@ -26,24 +26,26 @@
 #include <inttypes.h>
 
 #include "ztimer.h"
+#include "ztimer/periph_ptp.h"
 #include "ztimer/periodic.h"
 #include "fmt.h"
 #include "mutex.h"
 
 static mutex_t _mutex = MUTEX_INIT_LOCKED;
 
-#define MAX_OFFSET 2
+#define MAX_OFFSET 10000
 #define N 5
-#define INTERVAL 100LU
+#define INTERVAL 10000000LU
 
 static uint32_t _times[N];
+static ztimer_periph_ptp_t _clock;
 
 static int callback(void *arg)
 {
     (void)arg;
     static int count = 0;
 
-    _times[count] = ztimer_now(ZTIMER_MSEC);
+    _times[count] = ztimer_now(&_clock);
 
     count += 1;
 
@@ -64,14 +66,15 @@ static int callback(void *arg)
 int main(void)
 {
     ztimer_periodic_t t;
+    ztimer_periph_ptp_init(&_clock);
 
-    ztimer_periodic_init(ZTIMER_MSEC, &t, callback, NULL, INTERVAL);
 
-    ztimer_periodic_init(ZTIMER_MSEC, &t, callback, NULL, INTERVAL);
-    uint32_t last = ztimer_now(ZTIMER_MSEC);
+    ztimer_periodic_init(&_clock, &t, callback, NULL, INTERVAL);
+    uint32_t last = ztimer_now(&_clock);
 
     ztimer_periodic_start(&t);
 
-    if (!ztimer_is_set(ZTIMER_MSEC, &t.timer)) {
-        print_str("Test failed\n");
+    if (!ztimer_is_set(&_clock, &t.timer)) {
+        print_str("Test failed: ztimer_is_set()\n");
         return 1;
     }
 
```

**Comment by maribu** [comment]
Feel free to rebase on top of https://github.com/RIOT-OS/RIOT/pull/16221 to ease testing.

**Comment by maribu** [comment]
Ah, again almost forgot: Tge KConfig modeling should also be done

**Comment by Jnae** [comment]
I rebased it on #16221. And I also changed the KConfig file.

**Comment by maribu** [comment]
Two unrelated commits seem to have sneaked in. Could you run `git rebase master -i` and drop the two unrelated commits by removing the corresponding lines?

**Comment by Jnae** [comment]
> Two unrelated commits seem to have sneaked in. Could you run `git rebase master -i` and drop the two unrelated commits by removing the corresponding lines?

sorry, don't know what happend there. I fixed it

**Review by maribu — APPROVED** [review]
ACK. We still need to wait for https://github.com/RIOT-OS/RIOT/pull/16221 to get in, though.

Thanks for adding this! :-)

**Comment by maribu** [comment]
Needs a rebase on `master` now :-)

**Comment by maribu** [comment]
Can you rebaee again on the current state of master? Somehow, the bugfix of #16221 is still part of this PR

**Comment by maribu** [comment]
Thanks! This was an excellent contribution :-)

**Comment by kfessel** [comment]
maybe a warning should be added that this does a full (32 Bit) timer cylcle multiple times a day

**Comment by maribu** [comment]
I think @kaspar030 wanted to have (optional) 64 bit support in ztimer anyway. @Jnae is already working on allowing to pass through 64 bit hardware timestamps when `ztimer_now64` is used.

But in the meantime, we can have that :-)
